pkg/kv/kvserver/rangefeed/scheduler.go

```diff
@@ -118,55 +129,96 @@ type SchedulerConfig struct {
 // Each event is represented as a bit mask and multiple pending events could be
 // ORed together before being delivered to processor.
 type Scheduler struct {
-	SchedulerConfig
-
-	mu struct {
-		syncutil.Mutex
-		nextID int64
-		procs  map[int64]Callback
-		status map[int64]processorEventType
-		queue  *idQueue
-		// No more new registrations allowed. Workers are winding down.
-		quiescing bool
-	}
-	cond *sync.Cond
-	wg   sync.WaitGroup
+	nextID atomic.Int64
+	shards []*schedulerShard // id % len(shards)
+	wg     sync.WaitGroup
 }
 
 // NewScheduler will instantiate an idle scheduler based on provided config.
 // Scheduler needs to be started to become operational.
 func NewScheduler(cfg SchedulerConfig) *Scheduler {
-	if cfg.BulkChunkSize == 0 {
-		cfg.BulkChunkSize = enqueueBulkMaxChunk
+	bulkChunkSize := cfg.BulkChunkSize
+	if bulkChunkSize == 0 {
+		bulkChunkSize = enqueueBulkMaxChunk
 	}
-	s := &Scheduler{
-		SchedulerConfig: cfg,
-		wg:              sync.WaitGroup{},
+
+	s := &Scheduler{}
+
+	// Create shards.
+	numShards := 1
+	if cfg.ShardSize > 0 && cfg.Workers > cfg.ShardSize {
+		numShards = (cfg.Workers-1)/cfg.ShardSize + 1 // ceiling division
 	}
-	s.mu.procs = make(map[int64]Callback)
-	s.mu.status = make(map[int64]processorEventType)
-	s.mu.queue = newIDQueue()
-	s.cond = sync.NewCond(&s.mu)
+	for i := 0; i < numShards; i++ {
+		shardWorkers := cfg.Workers / numShards
+		if i < cfg.Workers%numShards { // distribute remainder
+			shardWorkers++
+		}
+		if shardWorkers <= 0 {
```

Comment:
In some sense, the same argument applies for 2 shards -- if they're slow, the third shard will struggle. As long as we have a worker, we should make forward progress, so I think this is fine (in particular since it requires the user to explicitly configure it this way).